src/sentry/utils/rollout.py

```diff
@@ -0,0 +1,152 @@
+      6. Complete your migration, secure in your knowledge that it's safe to do so.
+      7. Clean up your control branch & SafeRolloutComparator when you're done. Success!
+
+    Used like:
+    ```
+    callsite = "BarClass::baz"
+    control_data = old_slow_trustworthy_method()
+    if FooComparator.should_check_experiment(callsite):
+        experimental_data = new_fast_risky_method()
+        data = FooComparator.check_and_choose(
+            control_data,
+            experimental_data,
+            callsite,
+            len(experimental_data) == 0,
+            lambda ctl, exp: exp.issubset(ctl)
+        )
+    else:
+        data = control_data
+    ```
+    """
+
+    # This is the high-level eval rollout option. If this option is disabled, the
+    # should_check_experiment function will return False. Should be registered like:
+    # register(
+    #   "rollouts.your_rollout_name_here.should_eval_experimental",
+    #   type=Bool,
+    #   default=False,
+    #   flags=FLAG_MODIFIABLE_BOOL | FLAG_AUTOMATOR_MODIFIABLE,
+    # )
+    EVAL_EXPERIMENTAL_ROLLOUT_OPTION: str
+
+    # This is the callsite-level eval rollout option. If the option contains a callsite,
+    # the should_check_experiment function will return False. (This is useful if you see
+    # one callsite in particular start throwing.) Should be registered like:
+    # register(
+    #   "rollouts.your_rollout_name_here.eval_callsite_blocklist",
+    #   type=Sequence,
+    #   default=[],
+    #   flags=FLAG_ALLOW_EMPTY | FLAG_AUTOMATOR_MODIFIABLE,
+    # )
+    EVAL_EXPERIMENTAL_DATA_CALLSITE_BLOCKLIST_OPTION: str
+
+    # This is the callsite-level use-experimental-path rollout option. If the option
+    # contains a callsite, then that callsite will use the experimental-path data.
+    # This should generally only be used once you've determined that there is a high
+    # rate of partial- or exact- match at the callsite. Should be registered like:
+    # register(
+    #   "rollouts.your_rollout_name_here.eval_callsite_blocklist",
+    #   type=Sequence,
+    #   default=[],
+    #   flags=FLAG_ALLOW_EMPTY | FLAG_AUTOMATOR_MODIFIABLE,
+    # )
+    USE_EXPERIMENTAL_DATA_CALLSITE_ALLOWLIST_OPTION: str
+
+    # This is your rollout, which you can filter the DataDog dashboards to show.
+    # TODO @cpaul: construct DataDog dashboards once you have a rollout using this.
+    ROLLOUT_NAME: str
+
+    @classmethod
+    def should_check_experiment(cls, callsite: str) -> bool:
+        """
+        This function should control whether you evaluate your experimental branch at
+        all. Useful for rolling out by region or blocklisting callsites that throw.
+        """
+        if not options.get(cls.EVAL_EXPERIMENTAL_ROLLOUT_OPTION):
+            return False
+
+        return callsite not in options.get(cls.EVAL_EXPERIMENTAL_DATA_CALLSITE_BLOCKLIST_OPTION)
+
+    @classmethod
+    def check_and_choose(
```

Comment:
Definitely out of scope for this PR, but just thinking that eventually we might want this function to accept thunks instead of the evaluated values (at least for our use case of occurrences on EAP), since double reads will ofc add latency and be expensive if we do them 100% of the time. So maybe eventually we want to change this to a sampling approach, but it should be easy to make that change later on.